PR #1 to merge testing to main

umangS77 opened the pull request.

balamma committed:
added content

umangS77 committed:
Adding files

Commit:
Merge pull request #1 from umangS77/dev

Adding files

Commit:
Merge pull request #2 from umangS77/testing

Testing

Commit:
Testing a change in vw

Commit:
Merge pull request #3 from umangS77/testing

Testing a change in vw

Commit:
checking vmin instead of vw

Commit:
Checking vmin instead of vw

Commit:
vmin falied | testing rem

Commit:
vmin falied | testing rem

Commit:
rem falied | testing em

Commit:
rem failed | testing em

Commit:
Trying container height = 60px with rem units

Commit:
Trying container height = 60px with rem units

Commit:
Changes in index to account for missing images

Commit:
Merge pull request #4 from virtual-labs/testing

Merge testing to dev

Commit:
Removing Z coordinates

Commit:
Updating css to check iframe

Commit:
Adjusting the width of experiment container

Commit:
commented width

Commit:
Edited container height to 90px

Commit:
checking container height 120px

Commit:
Testing vmin as unit

Commit:
Testing vmin as unit

Commit:
testing rem

Commit:
testing rem

Commit:
testing vw

Commit:
testing vw

Commit:
testing vh

Commit:
testing vh

Commit:
testing rem again - vw gets cropped

Commit:
testing rem - vw gets cropped

Commit:
editing container css

Commit:
trying vw again

vw seems the best option available. rem crops the frame from sides and responsiveness stops working. 
vw crops the frame from bottom. Gotta ask the reason

Commit:
trying vw again

pavanchow committed:
Update deployment-script.yml

pavanchow committed:
Merge pull request #5 from virtual-labs/dev

Dev

Commit:
Corrected js file

Commit:
Fixed reset button issue

Commit:
Fixed reset button issue

Commit:
Added pretest and posttest

Commit:
Updated index

Commit:
Updated css

Commit:
Updated js

Commit:
Add files via upload

Commit:
Fixed phone screen button issue

raj-vlabs commented on experiment/posttest.json:
```diff
@@ -0,0 +1 @@
+[{"question":"1.This is a sample question:","answers":{"a":"This is a sample answer A","b":"This is a sample answer B"},"correctAnswer":"a"},{"question":"<img src='images/8.PNG'><br>Identify the location of Secondary electron detector","answers":{"a":"<img src='images/1b.png'>","b":"<img src='images/1a.png'>","c":"<img src='images/1c.PNG'>"},"correctAnswer":"c"}]
```
Posttest has sample questions. Please remove them and add real questions if available, otherwise leave an empty array and contact SME for questions.

raj-vlabs commented on experiment/pretest.json:
```diff
@@ -0,0 +1 @@
+[{"question":"1.This is a sample question:","answers":{"a":"This is a sample answer A","b":"This is a sample answer B"},"correctAnswer":"a"},{"question":"<img src='images/8.PNG'><br>Identify the location of Secondary electron detector","answers":{"a":"<img src='images/1b.png'>","b":"<img src='images/1a.png'>","c":"<img src='images/1c.PNG'>"},"correctAnswer":"c"}]
```
Pretest has sample questions. Please remove them and add real questions if available, otherwise leave an empty array and contact SME for questions.

umangS77 replied:
Solved - Empty array added

raj-vlabs commented on experiment/simulation/css/main.css:
```diff
@@ -1 +1,208 @@
+    }
+}
+
+
```
Please try to use classes instead of ids for any css that has a potential to be shared. Only the css that will always be specific to an element should be based on ids. This should consider the cases where the experiment can be enhanced or the css needs to be taken out and wrapped in a library for use with multiple experiments.

umangS77 replied:
Done

raj-vlabs commented on experiment/simulation/index.html:
```diff
@@ -1,12 +1,48 @@
+
 </head>
 <body>
 		<!-- Your code goes here-->
```
Please remove this comment.

umangS77 replied:
Removed

raj-vlabs commented on experiment/simulation/index.html:
```diff
@@ -1,12 +1,48 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
 	<!-- Add CSS at the head of HTML file -->
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/animejs/3.2.1/anime.min.js" integrity="sha512-z4OUqw38qNLpn1libAN9BsoDx6nbNFio5lA6CuTp9NlK83b89hgyCVq+N5FdBJptINztxn1Z3SaKSKUS5UP60Q==" crossorigin="anonymous"></script>
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
```
Bootstrap3 is not a recommended library because it is not responsive by default. Bootstrap4 and above is recommended. This is the reason why you had to write extra css for responsiveness. Please consider moving to bootstrap4.

umangS77 replied:
Changed to Bootstrap4

raj-vlabs commented on experiment/simulation/index.html:
```diff
@@ -1,12 +1,48 @@
+        <div onclick="stir()" id="rod"></div>
+        <button  class="btn-primary reset-btn" onclick="window.location.reload()">Reset Experiment</button>
+    </div>   
+    </div> 
```
Please fix the indentation of the file.

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
 //Your JavaScript goes in here
```
Please remove this comment.

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+        duration: 800,
+        easing: 'linear'
+    });
+    animation_1.add({
```
Do animation_1 ... animation_n signify the sequence of animations or they can be given more meaningful names?

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+
+function drop_pipette_to_beaker() {
+
+    console.log('pipette flag= ');
```
If these log messages do not need to be displayed in the prod application, can we use console.debug please?

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+let knob_heater_flag = 0; // turn-on or turn-off the heater
+let pipette_flag = 0; // handles the movement of pipette
+
+function pour( content ){
```
Where is the content variable used? Did you mean to pass top-beaker to the function instead of hard-coding the value on line 26? Only one of them is needed here, please remove the other one.

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+
+function pour( content ){
+    const image = document.getElementById("top-beaker");
+    image.style.pointerEvents = 'none';
```
Why is it required to block the pointer events?

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+    "  Click on the Pipette again to take 10ml of conc. FeCl3 solution.  ", 
+    "  Click on the Pipette to add the FeCl3 to the hot water in conical flask.  ",
+    "  Click on the Pipette again to add the FeCl3 drop by drop into the conical flask.  ", 
+    "  Fe(OH)3 colloid is obtained...  "];
```
It's good that you have used an array to store all messages in one place. It'd be better to use a map with a meaningful key so for example messages['POUR_WATER'] would be more readable than messages[0].

umangS77 replied:
Explained the reasons.

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+        delay: '800',
+        rotateZ: '-90'
+    }).add({
+        update: function(anim) {
```
Why are we passing anim to all update functions when they are not used in any of them?

umangS77 replied:
Resolved

raj-vlabs replied:
I still see the anim argument on line 44 of main.js. Please remove it.

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+
+    console.log('pipette flag= ');
+    console.log(pipette_flag);
+    if(pipette_flag === 0)
```
Can you use an enum to describe the different states of the pipette_flag instead of using numbers? enums are not supported natively in javascript but there are good alternatives available like using a map.

umangS77 replied:
Added comments and explanations

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+                document.getElementById("message").innerHTML = message;
+            }
+        });
+        pipette_flag = 2;
```
There's no transition for pipette_flag value of 1. From 0 it jumps to 2.

umangS77 replied:
Edited flags

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,416 @@
+    }
+}
+
+function pipette_fill() {
```
It'd be better to use one encapsulating function for all pipette transitions. This function will have a switch for the pipette_flag and then will call the respective functions based on the pipette_flag value. This way:
1. All transitions will be in one place for anyone to see
2. Functions will not need individual checks for flag value and increments, this stuff can be dealt with in the switch case. That way these functions will be independent of that control code and will focus on their functionality

umangS77 replied:
Added comments and explanations

raj-vlabs reviewed (changes requested):
1. objective.md is missing from the repo. Is it not required for this experiment?
2. Pretest and posttest have sample questions.
3. Prefer class based styling over id based styling.
4.  Remove guide comments.
5. Bootstrap4 is recommended, bootstrap3 is not.
6. simulation/index.html file not indented correctly.
7. Use console.debug for debugging messages.
8. Remove unused variables.
9. Multiple suggestions for refactoring the code.

Commit:
Fixed indentation

Commit:
Updated css

Commit:
Fixed merge issues

Commit:
Added pipette_flag guide

umangS77 commented:
Sincere apologies for the issues faced. I have tried my best to correct them now.

1. objective.md is missing from the repo. Is it not required for this experiment?
Sir the objective of the experiment is same as the aim of the experiment and aim.md file was already present.

2. Pretest and posttest have sample questions.
Sir the questions for pretest and post test as given in assignment.md are not option based. Thats why I did not add the questions to pretest and posttest files.

3. Prefer class based styling over id based styling.
Yes sir I understand that. Thats why I'e used ids for components like spoon, solution, beaker etc. and classes for components like reset-btn and title.

4. Remove guide comments.
Removed.

5. Bootstrap4 is recommended, bootstrap3 is not.
Changed to Bootstrap 4

6. simulation/index.html file not indented correctly.
Corrected the indentation

7. Use console.debug for debugging messages.
Removed debugging messages

8. Remove unused variables.
Removed

9. Concerns about pipette flag
pipette_flag variable signifies the various movements of pipette from one position to another.
Since clicking on one item invokes multiple functions, I have used flags as a system to direct the operations.
Flags signify which part of the function has to be executed when the corresponding component is clicked for nth time.
I have added the comment to explain flags.

10. Do animation_1 ... animation_n signify the sequence of animations or they can be given more meaningful names?
Yes. The code is written according to the sequence of motion of components. That's why I thought numbering the animation sequences would be better to understand.

11. Why is it required to block the pointer events?
Could you please elaborate this issue?

12. It's good that you have used an array to store all messages in one place. It'd be better to use a map with a meaningful key so for example messages['POUR_WATER'] would be more readable than messages[0].
I understand that sir. But the instruction are stored in the sequences of animation. Maps can be used but then the "sequence of instructions" would lose its meaning. Anyone trying to figure out the sequence will have to match each key for the sequence which I think would not be an ideal solution.
Still, if you think it is the correct way,please reply and I'll try to store it in format of Maps.

13. Why are we passing anim to all update functions when they are not used in any of them?
That was some testing component from my side.
Removed.

14. There's no transition for pipette_flag value of 1. From 0 it jumps to 2.
Error from my side.
Removed.

15. It'd be better to use one encapsulating function for all pipette transitions. This function will have a switch for the pipette_flag and then will call the respective functions based on the pipette_flag value.
I understand that sir. The pipette flag does the same function.
	-"All transitions will be in one place for anyone to see."
	I have used comments for the same
	-"Functions will not need individual checks for flag value and increments, this stuff can be dealt with in the switch case. That way these functions will be independent of that control code and will focus on their functionality"
	We still will need to check for flags since clicking each component invokes multiple functions.(Explained Above)


I hope this clears some of the merge issue. Further suggestions are most welcome. 
Regards,
Umang Srivastava

raj-vlabs commented on experiment/simulation/index.html:
```diff
@@ -1,13 +1,44 @@
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+        <script src="https://cdnjs.cloudflare.com/ajax/libs/animejs/3.2.1/anime.min.js" integrity="sha512-z4OUqw38qNLpn1libAN9BsoDx6nbNFio5lA6CuTp9NlK83b89hgyCVq+N5FdBJptINztxn1Z3SaKSKUS5UP60Q==" crossorigin="anonymous"></script>
+        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
```
I see that you have updated the bootstrap version. Do you still need to continue with the bootstrap.css and jquery versions, since they are both dependencies of bootstrap?

umangS77 replied:
Removed dependencies

raj-vlabs reviewed (commented):
1. Check whether bootstrap.css and jquery versions also need to be updated since you updated bootstrap.js version.

Commit:
Removed jquery and bootstrap3.0

raj-vlabs commented on experiment/simulation/index.html:
```diff
@@ -1,13 +1,44 @@
+                <div class = "row">
+                    <img id="bench" class = "bench" src="./images/bench.png" alt="Image">
+                    <div class = "water" id="water"></div>
+                    <img onclick="pour('top-beaker')" id="top-beaker" src="./images/beaker.png" alt="Image">
```
You do not need to pass an argument to pour function here since you removed the parameter.

umangS77 replied:
Removed

Commit:
Updated main.js

Commit:
Removed sample question

Commit:
Removed sample question

raj-vlabs commented on experiment/simulation/js/main.js:
```diff
@@ -1 +1,411 @@
+        duration: 400,
+        easing: 'linear',
+    }).add({
+        opacity: 0.4,
```
opacity is defined as a string everywhere else, if both ways are correct, choose one for consistency

umangS77 replied:
Changed datatype to string everywhere for consistency

raj-vlabs reviewed (changes requested):
1. main.js:Line 46 parameter anim still present on the update function.
2. opacity value is given as string at a few places and as a number at others. Choose the correct format. If both representations work, choose anyone for consistency.

Commit:
Edited pour func - solved issue

Commit:
Changed opacity datatype

raj-vlabs reviewed (changes requested):
main.js:line 44 still has parameter anim in the update function

Commit:
Removed anim
